### PR TITLE
LIME-131 Align passport-back template to enable deploying to dev-platform

### PR DESCRIPTION
**Pull request by smsgds**
### What changed

- Changed the creation of resources to be under the stack name instead of the environment name. /${ AWS_STACK_NAME }/resource vs /${ENVIRONMENT}/resource
- Secret parameters (keys/certs) values are to placed manually under /${ENVIRONMENT}/parameterpath
- Template created parameters are placed under /${ AWS_STACK_NAME }/parameterpath
- Relocated some resources in the template to be roughly the same location as the other CRIs.
- Added common cri environment parameters and conditions. 
- Added a mapping for ProvisionedConcurrency.
- Added MemorySizeMapping.
- Added mapping for MaximumAttemptCount (2 in all environments)
- Added mapping for AuthCodeExpirySeconds (600 in all environments)
- Set Lambda architecture globally (arm64). 
- Removed the prefixes from resource names e.g IPVCRIPassportBackPrivateAPI to PrivateUKPassportAPI
- DB Table names changed to the format !Sub "tablename-${AWS::StackName}" 
- Added parameters to record the table-names e.g /${AWS::StackName}/ParameterSessionTableName
- Switched to using Ref Environment for stagenames.
- Switched over to using the CRI-VPC - needs tested when in a build promoted to environment where VPC is required.
- API keys have been switched to !ImportValue core-infrastructure-ApiKey1 /ApiKey2.
- Add the deploy script but changed to also call the buildZip task to create the deployable zip files.
- - Code uris updated to point to where these files are created.
- The base url of the PublicAPI and PrivateAPI are exported so they can be used with IPV-Core stub/Passport front running locally.
- Private API is set to regional, and permissions set to allow when the environment is dev - so the stack can be reached when deployed on AWS.
- All the current lambdas updated to find resources from either the deployed stack or an environment set variable (if a secret) - avoids every new dev stack needing to set keys and certs
	- HTTPClient secrets located under "${Environment}/credentialIssuers/ukpassport/HttpClient/*"
	- DCS secrets under "${Environment}/credentialIssuers/ukpassport/DCS/*".
- getSsmParameter changed to two methods one getting per enviroment values, the other per stack.
	- The method duplicated for this is intended to be latter to be replaced with the cri-lib method that accepts a path prefix - e.g environment/stackname.
- getEncryptedSsmParameter renamed to getEncryptedEnvironmentSsmParameter and made private (only used internally in the same class), for naming consistency.
- updated .gitignore to ignore user generated samconfig.toml and .aws-sam folder in guided config creation.
- updated readme with deployment instructions.

### Why did it change

These template changes are the first step in aligning the passport CRI.
The objective was to allow passport-back to deploy on the dev-platform in the a method that is as close as possible to other CRI's with the minium code changes nessasary.

### Issue tracking
- [LIME-131](https://govukverify.atlassian.net/browse/LIME-131)
- [LIME-151](https://govukverify.atlassian.net/browse/LIME-151)
- [LIME-109](https://govukverify.atlassian.net/browse/LIME-109)

### CODE REVIEW

File: infrastructure/lambda/template.yaml [hunk 1]
```diff
@@ -0,0 +1,1138 @@
...
+  IPVCoreStubRedirectURIMapping:
+    Environment:
+      dev: "http://localhost:8085/callback"
+      build: "https://di-ipv-core-stub.london.cloudapps.digital/callback"
+      staging: "https://di-ipv-core-stub.london.cloudapps.digital/callback"
+      integration: "https://di-ipv-core-stub.london.cloudapps.digital/callback"
+
+  IPVCore1RedirectURIMapping:
+    Environment:
+      staging: "https://identity.staging.account.gov.uk/credential-issuer/callback?id=passport"
```
Comment (dannyhayes-gds) [inline infrastructure/lambda/template.yaml:181]:
```suggestion
      staging: "https://identity.staging.account.gov.uk/credential-issuer/callback?id=ukPassport"
```

File: infrastructure/lambda/template.yaml [hunk 2]
```diff
@@ -0,0 +1,1138 @@
...
+      integration: "https://identity.integration.account.gov.uk/credential-issuer/callback?id=passport"
```
Comment (dannyhayes-gds) [inline infrastructure/lambda/template.yaml:182]:
```suggestion
      integration: "https://identity.integration.account.gov.uk/credential-issuer/callback?id=ukPassport"
```

File: infrastructure/lambda/template.yaml [hunk 3]
```diff
@@ -0,0 +1,1138 @@
...
+      production: "https://identity.account.gov.uk/credential-issuer/callback?id=passport"
```
Comment (dannyhayes-gds) [inline infrastructure/lambda/template.yaml:183]:
```suggestion
      production: "https://identity.account.gov.uk/credential-issuer/callback?id=ukPassport"
```

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/config/ConfigurationVariable.java [hunk 4]
```diff
@@ -1,29 +1,26 @@
...
-    DCS_TLS_ROOT_CERT("/%s/credentialIssuers/ukPassport/dcs/tlsRootCertificate"),
-    JAR_ENCRYPTION_KEY_ID("/%s/credentialIssuers/ukPassport/self/jarKmsEncryptionKeyId"),
-    JAR_KMS_PUBLIC_KEY("/%s/credentialIssuers/ukPassport/self/jarKmsEncryptionPublicKey"),
-    MAX_JWT_TTL("/%s/credentialIssuers/ukPassport/self/maxJwtTtl"),
-    MAXIMUM_ATTEMPT_COUNT("/%s/credentialIssuers/ukPassport/self/maximumAttemptCount"),
-    PASSPORT_CRI_CLIENT_AUDIENCE("/%s/credentialIssuers/ukPassport/self/audienceForClients"),
-    PASSPORT_CRI_CLIENT_AUTH_MAX_TTL("/%s/credentialIssuers/ukPassport/self/maxJwtTtl"),
+    AUTH_CODE_EXPIRY_CODE_SECONDS("/%s/AuthCodeExpirySeconds"),
+    SESSION_TTL("/%s/SessionTtl"),
+    DCS_POST_URL_PARAM("/%s/credentialIssuers/ukpassport/DCS/PostUrl"),
```
Comment (dannyhayes-gds) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/config/ConfigurationVariable.java:6]: I'm guessing some of these changes means your going to update this in our config repo as well?

I think there are a few ssm params here that aren't being created in the new template file, so the new param names will need adding in our config repo.

Comment (gds-jbentham) [reply lib/src/main/java/uk/gov/di/ipv/cri/passport/library/config/ConfigurationVariable.java:6]: @smsgds this is one we will need to pickup theres a thread about it in the CRI LIME chat

**Review by gds-jbentham — APPROVED** [review]

**Pull request merged**